tools/log_processing/ems_refresh_timings.rb

```diff
@@ -0,0 +1,146 @@
+    opt :ems, 'Filter by a specific ems, comma separted', :type => :string
+  end
+
+  options[:sort_by] = opts[:sort_by].to_sym
+  [:target, :target_type, :ems].each do |key|
+    options[key] = opts[key].split(',') unless opts[key].nil?
+  end
+
+  argv.each do |arg|
+    logfiles << File.expand_path(arg) if File.file?(arg)
+  end
+
+  # If no logs were given on command line default to main project evm.log
+  logfiles << File.join(RAILS_ROOT, 'log/evm.log') if logfiles.empty?
+
+  [options, logfiles]
+end
+
+def filter(hash, opts = {})
+  return false unless opts[:target_type].nil? || opts[:target_type].include?(hash[:target_type])
+  return false unless opts[:target].nil? || opts[:target].include?(hash[:target])
+  return false unless opts[:ems].nil? || opts[:ems].include?(hash[:ems])
+  true
+end
+
+def parse_refresh_target(line)
+  if line =~ /EMS:? \[(.*?)\].+Refreshing target ([^\s]+).\[(.*?)\].+Complete/
+    {
+      :time        => line.time,
+      :ems         => Regexp.last_match[1],
+      :target_type => Regexp.last_match[2],
+      :target      => Regexp.last_match[3]
+    }
+  end
+end
+
+def parse_refresh_timings(line, targets)
+  if line =~ /EMS:? \[(.*?)\].+Refreshing targets for EMS...Complete - Timings:? (\{.+)$/
+    ems             = Regexp.last_match[1]
+    # Refresh timings are printed to the log as a hash, just eval it
+    refresh_timings = eval(Regexp.last_match[2])
+
+    # Find the most recent refresh target for our ems, since there is
+    # only one refresh worker this "has to be" the right one
+    # If this changes in the future we'll have to add a PID lookup here
+    refresh_target  = targets[ems].last
+    refresh_target = {
+      :target      => "unknown",
+      :target_type => "unknown"
+    } if refresh_target.nil?
+
+    # Add other useful information to the refresh timings
+    refresh_timings[:ems]         = ems
+    refresh_timings[:end_time]    = Time.parse(line.time + ' UTC').utc
+    refresh_timings[:start_time]  = refresh_timings[:end_time] - refresh_timings[:total_time]
+    refresh_timings[:target]      = refresh_target[:target]
+    refresh_timings[:target_type] = refresh_target[:target_type]
+
+    refresh_timings
+  end
+end
+
+def sort_timings(timings, sort_key)
+  timings.sort_by { |t| t[sort_key] }
+end
+
+def print_results(all_timings, ems_timings, opts)
+  columns        = [:start_time, :end_time, :total_time, :ems, :target_type, :target]
+  column_lengths = [0, 0, 0, 0, 0, 0]
+
+  print "Found #{all_timings.length} refreshes from #{ems_timings.keys.length} providers\n"
+
+  # Calculate how much padding we need for each column
+  sort_timings(all_timings, opts[:sort_by]).each do |timing|
+    (0..columns.length - 1).each do |i|
+      column_lengths[i] = [column_lengths[i], timing[columns[i]].to_s.length].max
+    end
+  end
+
+  # Print the column headers
+  columns.each_with_index do |col, i|
+    print "#{col.to_s.ljust(column_lengths[i])}   "
+  end
```

Comment:
If you're just trying to print a tableized version, there is an extension methid we wrote to Array to do that: [Array#tableize](http://www.rubydoc.info/gems/more_core_extensions/MoreCoreExtensions/ArrayTableize#tableize-instance_method)

So this all simplifies to:

```
puts ems_timings.tableize(:columns => [:start_time, :end_time, :total_time, :ems, :target_type, :target])
```

You don't need the `:columns`, but the default column ordering is alphabetical...see the options for other ways to tweak the column ordering.